Document elastic indexed Jobs

danielvegamyhre opened the pull request:
Document Elastic Indexed Jobs which were added in https://github.com/kubernetes/kubernetes/pull/115236

sftim commented:
/retitle [WIP] Document elastic indexed Jobs

sftim commented:
/sig apps

sftim commented:
Should this PR target _dev-1.27_?

danielvegamyhre commented:
@alculquicondor this is ready for review

alculquicondor commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -837,6 +837,30 @@ checking if the Job has the annotation
+
+In v1.27 Elastic Indexed Jobs were introduced which allow scaling Indexed Jobs up/down
+by mutating both `.spec.Parallelism` and `.spec.Completions` together such that 
+`.spec.Paralellism` == `.spec.Completions`. Previously, the Batch API validation
```
```suggestion
`.spec.paralellism == .spec.completions`.```

alculquicondor commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -837,6 +837,30 @@ checking if the Job has the annotation
+It is enabled by default.
+{{< /note >}}
+
+In v1.27 Elastic Indexed Jobs were introduced which allow scaling Indexed Jobs up/down
```
```suggestion
From Kubernetes v1.27, Indexed Jobs can be scaled up/down
```

alculquicondor commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -837,6 +837,30 @@ checking if the Job has the annotation
+In v1.27 Elastic Indexed Jobs were introduced which allow scaling Indexed Jobs up/down
+by mutating both `.spec.Parallelism` and `.spec.Completions` together such that 
+`.spec.Paralellism` == `.spec.Completions`. Previously, the Batch API validation
+enforced the immutability of `.spec.Completions`, however, this validation was relaxed in
+1.27 to allow scaling Indexed Jobs up/down, subject to the constraints above.
```
We don't document past behavior, as users can see the documentation for older versions.

But maybe you can say that when `ElasticIndexedJob` feature is disabled, completions is immutable.

alculquicondor commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -837,6 +837,30 @@ checking if the Job has the annotation
+1.27 to allow scaling Indexed Jobs up/down, subject to the constraints above.
+
+There are use cases where a batch workload requires an autoscaled indexed job with stable
+DNS pod names, for example MPI, Horovord, Ray, PyTorch etc. Prior to 1.27 this was not
```
Remove everything starting "Prior to 1.27".

It's not explaining anything that wasn't mentioned already in the first paragraph.

soltysh replied:
Docs are always explaining the current state. If you want to describe alternative approach I'd suggest creating a new section describing the behavior when you disable `ElasticIndexedJob`, although I doubt that's super useful. A short info about how this works is sufficient, like Aldo mentioned earlier.

alculquicondor commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -837,6 +837,30 @@ checking if the Job has the annotation
+enforced the immutability of `.spec.Completions`, however, this validation was relaxed in
+1.27 to allow scaling Indexed Jobs up/down, subject to the constraints above.
+
+There are use cases where a batch workload requires an autoscaled indexed job with stable
```
```suggestion
There are use cases where a batch workload requires scaling an indexed Job
```

The static DNS names are a tangential feature. I wouldn't mention it here.

soltysh commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -837,6 +837,30 @@ checking if the Job has the annotation
+{{< feature-state for_k8s_version="v1.27" state="beta" >}}
+
+{{< note >}}
+In order to use this behavior, you must enable the `ElasticIndexedJob`
```
If the feature is on by default, then stating that you need to enable it is redundant, imo.

soltysh reviewed (commented):
Aldo's comment still apply

danielvegamyhre commented:
/milestone 1.27

divya-mohan0209 commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -837,6 +837,23 @@ checking if the Job has the annotation
+
+{{< feature-state for_k8s_version="v1.27" state="beta" >}}
+
+From Kubernetes v1.27, Indexed Jobs can be scaled up/down up/down
```
Is the repetition intended?
```suggestion
From Kubernetes v1.27, Indexed Jobs can be scaled up/down up/down
```

danielvegamyhre replied:
Removed repetition

divya-mohan0209 commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -837,6 +837,23 @@ checking if the Job has the annotation
+{{< feature-state for_k8s_version="v1.27" state="beta" >}}
+
+From Kubernetes v1.27, Indexed Jobs can be scaled up/down up/down
+by mutating both `.spec.pompletions` and `.spec.completions` together such that 
```
I'm assuming the first is spec.parallelism and we haven't introduced a new spec?

```suggestion
by mutating both `.spec.parallelism` and `.spec.completions` together such that 
```

tengqm commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -837,6 +837,23 @@ checking if the Job has the annotation
+From Kubernetes v1.27, Indexed Jobs can be scaled up/down up/down
+by mutating both `.spec.pompletions` and `.spec.completions` together such that 
+`.spec.parallelism == .spec.completions`. When the `ElasticIndexedJob`
+[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
```
The gate has been documented already?

danielvegamyhre replied:
Documented the `ElasticIndexedJob` feature gate and included it in this PR.

tengqm commented:
@danielvegamyhre Looks like you have shipped another commit to this PR, but the feedbacks from the reviewers are not yet addressed?

danielvegamyhre commented:
@tengqm I had clicked "Commit Suggestion" on one comment, but hadn't addressed the other comments yet. I have now addressed all the comments and this is ready for another look.

tengqm commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -837,6 +837,21 @@ checking if the Job has the annotation
+
+{{< feature-state for_k8s_version="v1.27" state="beta" >}}
+
+Indexed Jobs can be scaled up/down up/down by mutating both `.spec.parallelism` 
```
```suggestion
Indexed Jobs can be scaled up/down by mutating both `.spec.parallelism` 
```

tengqm commented on content/en/docs/reference/command-line-tools-reference/feature-gates.md:
```diff
@@ -94,6 +94,7 @@ For a reference to old feature gates that are removed, please refer to
 | `DownwardAPIHugePages` | `false` | Beta | 1.21 | 1.21 |
 | `DownwardAPIHugePages` | `true` | Beta | 1.22 | |
 | `DynamicResourceAllocation` | `false` | Alpha | 1.26 | |
+| `ElasticIndexedJob` | `true` | Beta` | 1.27 | |
```
Please add a short description for this feature at the end of the page, as we did for other features.

danielvegamyhre replied:
Done

danielvegamyhre commented:
@alculquicondor @tengqm Thanks for the reviews, I have addressed your comments and this is ready for another look.

mickeyboxell commented:
/lgtm

tengqm commented on content/en/docs/reference/command-line-tools-reference/feature-gates.md:
```diff
@@ -513,6 +514,7 @@ Each feature gate is designed for enabling/disabling a specific feature:
   so that validation, merging, and mutation can be tested without committing.
 - `DynamicResourceAllocation": Enables support for resources with custom parameters and a lifecycle
   that is independent of a Pod.
+- `ElasticIndexedJob`: Enables Indexed Jobs to be scaled up or down by mutating both `spec.completions` and `spec.parallelism` together such that `spec.completions == spec.parallelism`. See docs on [elastic Indexed Jobs](docs/concepts/workloads/controllers/job#elastic-indexed-jobs) for more details.
```
```suggestion
- `ElasticIndexedJob`: Enables Indexed Jobs to be scaled up or down by mutating both
  `spec.completions` and `spec.parallelism` together such that `spec.completions == spec.parallelism`.
  See docs on [elastic Indexed Jobs](docs/concepts/workloads/controllers/job#elastic-indexed-jobs)
  for more details.
```

tengqm commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -849,6 +849,18 @@ checking if the Job has the annotation
+
+You can scale Indexed Jobs up or down by mutating both `.spec.parallelism` 
+and `.spec.completions` together such that `.spec.parallelism == .spec.completions`. 
+When the `ElasticIndexedJob`[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) on the [API server](/docs/reference/command-line-tools-reference/kube-apiserver/)`
```
```suggestion
When the `ElasticIndexedJob`[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
on the [API server](/docs/reference/command-line-tools-reference/kube-apiserver/)`
```

danielvegamyhre replied:
Fixed

reylejano commented on content/en/docs/concepts/workloads/controllers/job.md:
```diff
@@ -849,6 +849,19 @@ checking if the Job has the annotation
 this annotation from Jobs. Instead, you can recreate the Jobs to ensure they
 are tracked using Pod finalizers.
 
+### Elastic Indexed Jobs
```
```suggestion
### Elastic indexed Jobs
```
Sentence case for headings, Jobs is capitalized because it's an API object

LukeMwila commented:
/lgtm

reylejano commented:
/approve
